src/store/builder/index.ts

```diff
@@ -113,27 +117,113 @@ export const mutations: MutationTree<IBuilderState> = {
     operations[operation.kind] = operation
   },
 
-  setBlockPositionTo(state, { position: { x, y }, block }) {
-    // todo: ensure our vendor_metadata.io_viamo is always instantiated with builder // uiData props
-    // if (!block.vendor_metadata.io_viamo.uiData) {
-    //   defaultsDeep(block, {vendor_metadata: {io_viamo: {uiData: {xPosition: 0, yPosition: 0}}}})
-    //   Vue.observable(block.vendor_metadata.io_viamo.uiData)
-    // }
-
-    block.vendor_metadata.io_viamo.uiData.xPosition = x
-    block.vendor_metadata.io_viamo.uiData.yPosition = y
-  },
-
   setIsEditable(state, value) {
     state.isEditable = value
   },
 
   initDraggableForExitsByUuid(state) {
     state.draggableForExitsByUuid = {}
+  },
+
+  updateBlockDraggablePosition(state, { uuid, position }: { uuid: IBlock['uuid'], position: IPosition }) {
+    Object.assign(state.draggableForBlocksByUuid[uuid], { left: position.x, top: position.y })
+  },
+
+  updateToolBarHeight(state, height) {
+    state.toolbarHeight = height
   }
 }
 
 export const actions: ActionTree<IBuilderState, IRootState> = {
+  setBlockPositionTo({ state, commit, getters, dispatch, rootState, rootGetters }, { position: { x, y }, block }) {
+    // todo: ensure our vendor_metadata.io_viamo is always instantiated with builder // uiData props
+    // if (!block.vendor_metadata.io_viamo.uiData) {
+    //   defaultsDeep(block, {vendor_metadata: {io_viamo: {uiData: {xPosition: 0, yPosition: 0}}}})
+    //   Vue.observable(block.vendor_metadata.io_viamo.uiData)
+    // }
+
+    if (!includes(rootState.flow.selectedBlockUuids, block.uuid)) {
+      block.vendor_metadata.io_viamo.uiData.xPosition = x
+      block.vendor_metadata.io_viamo.uiData.yPosition = y
+    } else { // the block is selected
+      // Store UI Position
+      const initialPosition = {
+        x: clone(block.vendor_metadata.io_viamo.uiData.xPosition),
+        y: clone(block.vendor_metadata.io_viamo.uiData.yPosition),
+      } as IPosition
+
+      // Prepare translation
+      const translationDelta: IPosition = {
+        x: x - initialPosition.x,
+        y: y - initialPosition.y
+      }
+
+      let shouldReversePosition = false;
+      if (translationDelta.x > 0 /* moving to the right */ ||
+        rootGetters['flow/selectedBlockAtTheFurthestLeftPosition'].vendor_metadata.io_viamo.uiData.xPosition > 0 /* left space still available */) {
+        block.vendor_metadata.io_viamo.uiData.xPosition = x
+      } else {
+        translationDelta.x = 0
+        block.vendor_metadata.io_viamo.uiData.xPosition = initialPosition.x
+        shouldReversePosition = true
+      }
+
+      if(translationDelta.y > 0 /* moving to the top */ ||
+        rootGetters['flow/selectedBlockAtTheTopPosition'].vendor_metadata.io_viamo.uiData.yPosition - state.toolbarHeight > 0 /* top space still available */) {
```

Comment:
This might get hard to maintain as the UI becomes more fancy. It feels fragile to try to match css and logic. Is there a way we can remove the need to track anything to do with toolbar height? My first thought is to use CSS to manage margins:

```
------------------------ Toolbar
------------------------ Alerts Toolbar
|----------------------| Canvas
|----------------------|
|----------------------|
|----------------------|
```

Using the same logic as we already use to show/hide alerts-toolbar, we can (create, then) apply `.has-alerts` class to canvas which can add a margin-top of whatever the magic number is. This mitigates mixing javascript and css, keeps pixel variations within css, and avoids having those magic numbers sprinkled throughout our formulas and vuex state. 

If we use margins instead of padding, then the top of what we can see on the canvas becomes the _true_ top of the canvas and the formulas needn't be concerned. 

What do you think about this?